odl/tomo/geometry/parallel.py

```diff
@@ -472,6 +472,42 @@ def __repr__(self):
         return '{}(\n{}\n)'.format(self.__class__.__name__,
                                    indent_rows(sig_str))
 
+    def __getitem__(self, slc):
+        """Return self[slc]
+
+        Examples
+        --------
+        Extract sub-geometry:
+
+        >>> apart = odl.uniform_partition(0, np.pi, 4)
+        >>> dpart = odl.uniform_partition(-1, 1, 20)
+        >>> geom = Parallel2dGeometry(apart, dpart)
+        >>> geom[::2, :]
```

Comment:
We need a discussion on what the expected behavior is here. I'm not totally convinced what we see here is correct.

Specifically I would also be OK with the `min_pt` and `max_pt` staying un-changed.

With that said, i guess any such change should go into `Partition.__getitem__`?

Reply:
Without reading any details, the basic expectation I have is that `geom[i]` gives the geometry for the i-th angle, leaving the rest unchanged. The handling of weightings is a discussion point, with three possibilities, namely (1) take cell volume, (2) take angle interval length, or (3) take 1.0.
I have a preference for (1) because it makes the product space operator with 1 angle per operator equivalent to the original operator.
Then, the remaining indices should index the detector, as defined by the partition.

Reply:
Looking at the example, I think that one should raise an error because we cannot "rip holes" into intervals. For a ray transform of each second angle, I think one should use `geoms = list(geometry)` and then `ops = [RayTransform(geom, ...) for geom in geoms]` and finally `op = BroadcastOperator(*ops)`.

I think there are some utilities like that in [this commit](https://github.com/kohr-h/odl/commit/fea1f92bd0bb7b2b0e6c279242f5ef95e9048e97).

Reply:
While i totally agree that `geometry[i]` should return the i:th angle so to say, weighting would then implicitly be (1).

Anyway regarding your approach of creating a single projector for each geometry, it simply is not performance-wise a feasible solution. For example, doing all at once (`'all'`) vs doing them all at once (`'single'`) for some typical geometry gives this timing:

                           all :      1.238 
                        single :     30.094
                        
So proper vectorization is a must, and that approach is not feasible.

Anyway, for me the only reasonable thing here is that

```python
geometry[...].partition == geometry.partition[...]
```

so I guess this discussion might as well be about what `partition[...]` should mean.

With that said, there are basically some options for what this code should do

```python
>>> partition = odl.uniform_partition(0, 10, 10)
>>> partition[::2]
```

* Current:
  ```python
  nonuniform_partition(
      [0.5, 2.5, 4.5, 6.5, 8.5],
      min_pt=0.0, max_pt=9.0
  )
  ```
  
* Keep full domain:
  ```python
  nonuniform_partition(
      [0.5, 2.5, 4.5, 6.5, 8.5],
      min_pt=0.0, max_pt=10.0
  )
  ```
  
* Centered:  
  ```python
  uniform_partition(0, 10, 5)
  ```
  
* Throw exception
  ```python
  IndexException(...)
  ```

I think I'd personally be in favor of the second option, e.g. keeping the far edges and selecting the inner points according to the indices.

Reply:
Perhaps @ozanoktem has some input on this?

Reply:
Well, the inefficiency is not completely our fault. Have you tried profiling the angle-by-angle code vs. bulk? I think I remember @wjp mention possible speedups in ASTRA for this case.

Anyway, assuming that we need a way of doing this in bulk, I'd definitely vote against option 3 since it moves grid points. That's a no-go IMO.

Regarding 1 vs. 2, I think we should look at the weights that the two suggest (keeping also `[0, 2*pi]` with periodicity in mind.
Let's check a 1D case where we partition an interval `[a, b]` with nodes `a <= x_1 < ... < x_n <= b`. When using nearest interpolation, the "inner" weights (`i = 2, ..., n-1`) are computed as 

    w_i = (x_{i+1} - x_{i-1}) / 2

For the left boundary `x_1` we get

    w_1 = (x_1 + x_2) / 2 - a

for the right boundary

    w_{n} = b - (x_{n-1} + x_n) / 2

If `a = x_1`, the left weight becomes `w_1 = (x_2 - x_1) / 2`, i.e., half the weight if the spacing in uniform.
Also, for periodic functions (`f_1 = f_n` implicitly, and `x_1 = a, x_n = b`) the combined boundary weight is

    w_1^* = (x_n - x_{n-1}) / 2 + (x_2 - x_1) / 2,

which is just `h/2 + h/2 = h` for uniform sampling.

Let's check that with the example above: 
```python
>>> partition = odl.uniform_partition(0, 10, 10)
>>> partition.grid.coord_vectors[0]  # grid points
array([ 0.5,  1.5,  2.5,  3.5,  4.5,  5.5,  6.5,  7.5,  8.5,  9.5])
>>> partition.cell_sizes_vecs[0]  # the w_i
array([ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])
```

Option 1:
```python
>>> subpart = odl.nonuniform_partition(
...     [0.5, 2.5, 4.5, 6.5, 8.5],
...     min_pt=0.0, max_pt=9.0
... )
>>> subpart.grid.coord_vectors[0]  # all good
array([ 0.5,  2.5,  4.5,  6.5,  8.5])
>>> subpart.cell_sizes_vecs[0]  # not quite as good
array([ 1.5,  2. ,  2. ,  2. ,  1.5])
```

Option 2:
```python
>>> subpart = odl.nonuniform_partition(
...     [0.5, 2.5, 4.5, 6.5, 8.5],
...     min_pt=0.0, max_pt=10.0
... )
>>> subpart.grid.coord_vectors[0]  # all good
array([ 0.5,  2.5,  4.5,  6.5,  8.5])
>>> subpart.cell_sizes_vecs[0]  # better
array([ 1.5,  2. ,  2. ,  2. ,  2.5])
```

Just for the records, also Option 3:
```python
>>> subpart = odl.uniform_partition(0, 10, 5)
>>> subpart.grid.coord_vectors[0]  # Nooooooo!
array([ 1.,  3.,  5.,  7.,  9.])
>>> subpart.cell_sizes_vecs[0]  # best, but worthless
array([ 2.,  2.,  2.,  2.,  2.])
```

Why is option 2 better? The boundary weights are not perfect, but at least they *sum up to the right value*, meaning that periodic functions will be weighted correctly.
There's little to be done about the doubling of the weights overall since the subintervals actually double in size. So a convenience function that, e.g., splits the `RayTransform` should take that weighting change into account. The goal should be that stacking together angles in an interleaved way should produce the same result after backprojecting as the original set of angles.

Reply:
I agree with what you write here and I think the conculsion (as per this PR) is that we go for option 2 for now (I'll update `partition.__getitem__`), specifically this means that we use the implementation as is for the geometries.